pyartcd/pyartcd/pipelines/build_rhcos.py

```diff
@@ -0,0 +1,197 @@
+        self.version = version
+        self.api_token = None
+
+        self.request_session = requests.Session()
+        retries = Retry(
+                total=5, backoff_factor=1,
+                status_forcelist=[500, 502, 503, 504],
+                method_whitelist=["HEAD", "GET", "POST"],
+        )
+        self.request_session.mount("https://", TimeoutHTTPAdapter(max_retries=retries))
+
+
+    def run(self):
+        self.request_session.headers.update({"Authorization": f"Bearer {self.retrieve_auth_token()}"})
+        current = self.query_existing_builds()
+        result = {}
+        if current and not self.ignore_running:
+            result["action"] = "skip"
+            result["builds"] = [
+                dict(url=b["url"], description=b["description"], result=None)
+                for b in current
+            ]
+        else:
+            # [lmeyer] there is a window of at least seconds after some other build is created,
+            # when the above check cannot see it, and so this may start a new one. interestingly,
+            # Jenkins seems to silently ignore identical builds in close succession (probably
+            # double-click protection - if the parameters differ both are started), and should this
+            # happen we will see a build start, assume we started it, and watch it to completion.
+            # this seems unlikely to cause any problems other than mild confusion.
+            self.start_build()
+            result["action"] = "build"
+            result["builds"] = self.wait_for_builds()
+
+        # final status in stdout
+        print(json.dumps(result))
+
+    def retrieve_auth_token(self) -> str:
+        """Retrieve the auth token from the Jenkins service account to use with Jenkins API"""
+        # use the first secret named after the jenkins service account (there can be several)
+        secret = next((s for s in oc.selector('secrets') if s.name().startswith('jenkins-token-')), None)
+        if secret is None:
+            raise Exception("Unable to find Jenkins service account token")
+
+        return base64.b64decode(secret.model.data.token).decode('utf-8')
+
+    @staticmethod
+    def build_parameters(build: Dict[str, List[Dict]]) -> Dict:
+        """Parse the build parameters from the build actions into a more usable dict"""
+        parameters = next((  # first action that has parameters
+            action["parameters"]
+            for action in build["actions"]
+            if "parameters" in action
+        ), [])
+        return {p["name"]: p["value"] for p in parameters}
+
+    @staticmethod
+    def build_url(job: str, number: int) -> str:
+        return f"{JENKINS_BASE_URL}/job/{job}/{number}/"
+
+    def query_existing_builds(self) -> List[Dict]:
+        """Check if there are any existing builds for the given version. Returns builds in progress."""
+        builds = []
+        for job in ("build", "build-arch", "release"):
+            builds.extend(
```

Comment:
Hmm so we are collecting all the in progress builds for all jobs and looking at those.
But what if for a build-id, 1 arch is in progress and we want to re-trigger another arch.
I guess we're not handling that case right now? This would also be easier if the main build job 
would wait on build-arch and release jobs..

Reply:
you could manually re-trigger another arch, and as long as it started before the others finished, this would adopt it as part of the build and wait on it.

but for the most part i don't want us re-triggering arch builds unless we're trying to get a release and watching them very closely (and ideally not at all).